homeassistant/components/threshold/binary_sensor.py

```diff
@@ -212,25 +212,37 @@ def above(threshold):
 
         if self.sensor_value is None:
             self._state_position = POSITION_UNKNOWN
-            self._state = False
+            self._state = None
+            return
 
-        elif self.threshold_type == TYPE_LOWER:
+        if self.threshold_type == TYPE_LOWER:
+            if self._state is None:
+                self._state = False
+                self._state_position = POSITION_ABOVE
             if below(self._threshold_lower):
                 self._state_position = POSITION_BELOW
                 self._state = True
             elif above(self._threshold_lower):
                 self._state_position = POSITION_ABOVE
                 self._state = False
+            return
 
-        elif self.threshold_type == TYPE_UPPER:
+        if self.threshold_type == TYPE_UPPER:
+            if self._state is None:
+                self._state = False
+                self._state_position = POSITION_BELOW
             if above(self._threshold_upper):
                 self._state_position = POSITION_ABOVE
                 self._state = True
             elif below(self._threshold_upper):
                 self._state_position = POSITION_BELOW
                 self._state = False
+            return
 
-        elif self.threshold_type == TYPE_RANGE:
+        if self.threshold_type == TYPE_RANGE:
+            if self._state is None:
+                self._state = True
+                self._state_position = POSITION_IN_RANGE
```

Comment:
It's really odd that the logic for the binary sensor is reversed when there's both a lower and an upper threshold.

Should we take the plunge and redefine it such that the threshold sensor is "on" when the monitored sensor is outside the range, in the same way as it's "on" if the monitored sensor is below a single lower threshold and on when it's above a single upper threshold?

Reply:
In the code block above, shouldn't the state depend on the current sensor value? This is going to be always true no matter what the current value is when it comes out of unknown, until the state changes again.

Reply:
Probably makes sense to just ignore the hysteresis in this context and base it purely on the current state being above/below/between the thresholds. We don't know the direction of sensor movement so it's our best guess. 

Other option would be something like (if / then so first one)
* sensor > upper + hysteresis -> Above
* sensor > upper - hysteresis -> Unknown
* sensor > lower + hysteresis -> Between
* sensor > lower - hysteresis > Unknown
* else: Below.

Reply:
> This is going to be always true no matter what the current value is when it comes out of unknown, until the state changes again.

No, that's not the case. We first check if the state of the binary sensor is unknown, in which case we prime it, and then we check the state of the source sensor.

Reply:
Oh yeh sorry my bad. Read the below if as an elif. Maybe a line break in there to make it clear?

Reply:
Between should be True IMHO. Reversing it isn't really what you'd expect.

Reply:
> Between should be True IMHO. Having it true outside isn't really what you'd expect.

A `binary_sensor` is "on" to signal the bad / abnormal state.
Consequently, a threshold sensor with only an upper threshold is `on` when the input sensor is above the limit and a threshold sensor with only a lower limit is `on` when the input sensor is below the limit.
It doesn't seem correct that this is reversed for a threshold sensor with both upper and lower limits; it should be `on` when the input sensor is outside the limits.

Reply:
Yeh, IDK. IMHO:
* Upper bound => value > upper -> True (which it what happens) 
* Lower bound => value < lower -> True
* Range => value between lower and upper -> True (what's happening now)

But I can see the argument for having it the other way. Not sure if that's enough to _breaking change_ it to the other way though. This PR is a breaking change to be sure, but a pretty soft one really in comparison to reversing the logic. in existing use.